package/endpoint_adapter/src/endpoint_adapter.c

```diff
@@ -698,287 +767,256 @@ static ssize_t handle_write_all_via_framer(handle_t *handle,
   return buffer_index;
 }
 
-static void io_loop_pubsub(handle_t *read_handle, handle_t *write_handle)
+static void io_loop_pubsub(pk_loop_t *loop, handle_t *read_handle, handle_t *write_handle)
 {
-  debug_printf("io loop begin\n");
+  if (read_handle->pk_ept != NULL) {
+    PK_METRICS_UPDATE(MR, MI.tx_read_count);
+  } else {
+    PK_METRICS_UPDATE(MR, MI.rx_read_count);
+  }
 
-  while (1) {
+  /* Read from read_handle */
+  static uint8_t buffer[READ_BUFFER_SIZE];
+  ssize_t read_count = handle_read(read_handle, buffer, sizeof(buffer));
+  if (read_count <= 0) {
+    debug_printf("read_count %d errno %s (%d)\n", read_count, strerror(errno), errno);
+    pk_loop_stop(loop);
+    return;
+  }
 
-    PK_METRICS_UPDATE(MR, MI.read_count);
+  if (read_handle->pk_ept != NULL) {
```

Comment:
Same here.

Reply:
fixed, added a helper macro since I forgot what the different between the 2 cases was